app/assets/javascripts/discourse/app/components/modal/bulk-topic-actions.gjs

```diff
@@ -0,0 +1,153 @@
+import Component from "@glimmer/component";
+import { action } from "@ember/object";
+import { inject as service } from "@ember/service";
+import { Promise } from "rsvp";
+import DButton from "discourse/components/d-button";
+import DModal from "discourse/components/d-modal";
+import Topic from "discourse/models/topic";
+import htmlSafe from "discourse-common/helpers/html-safe";
+import i18n from "discourse-common/helpers/i18n";
+//import AppendTags from "../bulk-actions/append-tags";
+//import ChangeCategory from "../bulk-actions/change-category";
+//import ChangeTags from "../bulk-actions/change-tags";
+//import NotificationLevel from "../bulk-actions/notification-level";
+
+export default class BulkTopicActions extends Component {
+  @service router;
+
+  async perform(operation) {
+    this.loading = true;
+
+    if (this.args.model.topics.length > 20) {
+      this.showProgress = true;
+    }
+
+    try {
+      return this._processChunks(operation);
+    } catch {
+      this.dialog.alert(i18n.t("generic_error"));
+    } finally {
+      this.loading = false;
+      this.processedTopicCount = 0;
+      this.showProgress = false;
+    }
+  }
+
+  _generateTopicChunks(allTopics) {
+    let startIndex = 0;
+    const chunkSize = 30;
+    const chunks = [];
+
+    while (startIndex < allTopics.length) {
+      const topics = allTopics.slice(startIndex, startIndex + chunkSize);
+      chunks.push(topics);
+      startIndex += chunkSize;
+    }
+
+    return chunks;
+  }
+
+  _processChunks(operation) {
+    const allTopics = this.args.model.topics;
+    const topicChunks = this._generateTopicChunks(allTopics);
+    const topicIds = [];
+
+    const tasks = topicChunks.map((topics) => async () => {
+      const result = await Topic.bulkOperation(topics, operation);
+      this.processedTopicCount = this.processedTopicCount + topics.length;
+      return result;
+    });
+
+    return new Promise((resolve, reject) => {
+      const resolveNextTask = async () => {
+        if (tasks.length === 0) {
+          const topics = topicIds.map((id) => allTopics.findBy("id", id));
+          return resolve(topics);
+        }
+
+        const task = tasks.shift();
+
+        try {
+          const result = await task();
+          if (result?.topic_ids) {
+            topicIds.push(...result.topic_ids);
+          }
+          resolveNextTask();
+        } catch {
+          reject();
+        }
+      };
+
+      resolveNextTask();
+    });
+  }
+
+  @action
+  setComponent(component) {
+    this.activeComponent = component;
+  }
+
+  @action
+  performAction() {
+    // TODO: return only visible topics
+    //visible: ({ topics }) => !topics.some((t) => t.isPrivateMessage),
```

Comment:
Ah I think this is just used in the old modal to control whether the Action button is visible. E.g. Archive Topics only shows when none of the topics are PMs:

https://github.com/discourse/discourse/blob/c01aa32a796a0d74d3ecfe34080ec6db1759802f/app/assets/javascripts/discourse/app/components/modal/topic-bulk-actions.js#L60-L68

So this `visible` is not necessary here, it'll be used in `TopicBulkSelectDropdown` instead.